Type inference regression

Servo compiles with rustc 71c06a56a120a0d5e3b224105ee3e6754f83e5fa 2016-12-18, but not with 439c3128d740af372dae163310f7292e999098e1 2016-12-21. Diff: https://github.com/rust-lang/rust/compare/71c06a56a...439c3128d7

The `style` crate fails to build with:

```rust
error[E0282]: unable to infer enough type information about `E`
  --> /home/simon/projects/servo/components/style/font_face.rs:65:9
   |
65 |         try!(self.family.to_css(dest));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `E`
   |
   = note: type annotations or generic parameter binding required
   = note: this error originates in a macro outside of the current crate
```

The `to_css` method being called here is:

```rust
impl ToCss for FontFamily {
    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
        // ...
    }
}
```

Its return type is `std::fmt::Result` which is `type Result = std::result::Result<(), std::fmt::Error>;`. This call is in another impl of the same method of the same trait:

```rust
impl ToCss for FontFaceRule {
    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
        try!(dest.write_str("@font-face { font-family: "));
        try!(self.family.to_css(dest));
        try!(dest.write_str(";"));

        // ...
    }
}
```

Adding an explicit type annotation works around this error, but only to uncover another very similar one, then another. The third one was in code generated by serde’s `derive(Serialize)`, so I gave up that approach.

```rust
        let r: fmt::Result = self.family.to_css(dest);
        try!(r);
```

So far I haven’t managed to reduce a test case smaller than "most of Servo", sorry.

I’m suspicious of 6ea1fbb52bc2f82a78d7b592bdfed2b05e6fb666 (which is part of #38099) because it’s commit message mentions inference. I’m not building a compiler #38099 to confirm.

Expanding the `try!` macro:

```diff
diff --git a/components/style/font_face.rs b/components/style/font_face.rs
index 70f904c17a..a3d0ca3882 100644
--- a/components/style/font_face.rs
+++ b/components/style/font_face.rs
@@ -62,7 +62,10 @@ impl ToCss for FontFaceRule {
     // Serialization of FontFaceRule is not specced.
     fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
         try!(dest.write_str("@font-face { font-family: "));
-        try!(self.family.to_css(dest));
+        match self.family.to_css(dest) {
+            Err(err) => return Err(err),
+            Ok(_) => {}
+        };
         try!(dest.write_str(";"));
 
         if self.sources.len() > 0 {
```

(Even simplifying it, this is not using `From`) gives:

```rust
error[E0282]: unable to infer enough type information about `T`
  --> /home/simon/projects/servo2/components/style/font_face.rs:66:13
   |
66 |             Err(err) => return Err(err),
   |             ^^^^^^^^ cannot infer type for `T`
   |
   = note: type annotations or generic parameter binding required

error: aborting due to previous error
```

Swapping the order of the two `match` arms:

```rust
error[E0282]: unable to infer enough type information about `T`
  --> /home/simon/projects/servo2/components/style/font_face.rs:66:13
   |
66 |             Ok(_) => {}
   |             ^^^^^ cannot infer type for `T`
   |
   = note: type annotations or generic parameter binding required

error: aborting due to previous error
```

Changing the `Ok(_)` pattern to `Ok(())`:

```rust
error[E0282]: unable to infer enough type information about `E`
  --> /home/simon/projects/servo2/components/style/font_face.rs:66:13
   |
66 |             Ok(()) => {}
   |             ^^^^^^ cannot infer type for `E`
   |
   = note: type annotations or generic parameter binding required

error: aborting due to previous error
```

Hacking `./mach build-geckolib` to not use the stable version of the compiler (so effectively disabling the `servo` feature and enabling the `gecko` one) also gives the same error.

--- a/src/librustc_resolve/resolve_imports.rs
+++ b/src/librustc_resolve/resolve_imports.rs
@@ -128,6 +128,7 @@ impl<'a> NameResolution<'a> {
 impl<'a> Resolver<'a> {
     fn resolution(&self, module: Module<'a>, ident: Ident, ns: Namespace)
                   -> &'a RefCell<NameResolution<'a>> {
+        let ident = ident.unhygienize();
         *module.resolutions.borrow_mut().entry((ident, ns))
                .or_insert_with(|| self.arenas.alloc_name_resolution())
     }
@@ -141,7 +142,6 @@ impl<'a> Resolver<'a> {
                                    ignore_unresolved_invocations: bool,
                                    record_used: Option<Span>)
                                    -> Result<&'a NameBinding<'a>, Determinacy> {
-        let ident = ident.unhygienize();
         self.populate_module_if_necessary(module);
 
         let resolution = self.resolution(module, ident, ns)
@@ -307,7 +307,6 @@ impl<'a> Resolver<'a> {
                       ns: Namespace,
                       binding: &'a NameBinding<'a>)
                       -> Result<(), &'a NameBinding<'a>> {
-        let ident = ident.unhygienize();
         self.update_resolution(module, ident, ns, |this, resolution| {
             if let Some(old_binding) = resolution.binding {
                 if binding.is_glob_import() {

--- a/src/libsyntax/fold.rs
+++ b/src/libsyntax/fold.rs
@@ -302,23 +302,22 @@ pub fn noop_fold_view_path<T: Folder>(view_path: P<ViewPath>, fld: &mut T) -> P<
     view_path.map(|Spanned {node, span}| Spanned {
         node: match node {
             ViewPathSimple(ident, path) => {
-                ViewPathSimple(ident, fld.fold_path(path))
+                ViewPathSimple(fld.fold_ident(ident), fld.fold_path(path))
             }
             ViewPathGlob(path) => {
                 ViewPathGlob(fld.fold_path(path))
             }
             ViewPathList(path, path_list_idents) => {
-                ViewPathList(fld.fold_path(path),
-                             path_list_idents.move_map(|path_list_ident| {
-                                Spanned {
-                                    node: PathListItem_ {
-                                        id: fld.new_id(path_list_ident.node.id),
-                                        rename: path_list_ident.node.rename,
-                                        name: path_list_ident.node.name,
-                                    },
-                                    span: fld.new_span(path_list_ident.span)
-                                }
-                             }))
+                let path = fld.fold_path(path);
+                let path_list_idents = path_list_idents.move_map(|path_list_ident| Spanned {
+                    node: PathListItem_ {
+                        id: fld.new_id(path_list_ident.node.id),
+                        rename: path_list_ident.node.rename.map(|ident| fld.fold_ident(ident)),
+                        name: fld.fold_ident(path_list_ident.node.name),
+                    },
+                    span: fld.new_span(path_list_ident.span)
+                });
+                ViewPathList(path, path_list_idents)
             }
         },
         span: fld.new_span(span)
@@ -345,7 +344,7 @@ pub fn noop_fold_arm<T: Folder>(Arm {attrs, pats, guard, body}: Arm, fld: &mut T
 pub fn noop_fold_ty_binding<T: Folder>(b: TypeBinding, fld: &mut T) -> TypeBinding {
     TypeBinding {
         id: fld.new_id(b.id),
-        ident: b.ident,
+        ident: fld.fold_ident(b.ident),
         ty: fld.fold_ty(b.ty),
         span: fld.new_span(b.span),
     }
@@ -673,7 +672,7 @@ pub fn noop_fold_ty_param<T: Folder>(tp: TyParam, fld: &mut T) -> TyParam {
             .collect::<Vec<_>>()
             .into(),
         id: fld.new_id(id),
-        ident: ident,
+        ident: fld.fold_ident(ident),
         bounds: fld.fold_bounds(bounds),
         default: default.map(|x| fld.fold_ty(x)),
         span: span
@@ -1088,7 +1087,7 @@ pub fn noop_fold_pat<T: Folder>(p: P<Pat>, folder: &mut T) -> P<Pat> {
                 let fs = fields.move_map(|f| {
                     Spanned { span: folder.new_span(f.span),
                               node: ast::FieldPat {
-                                  ident: f.node.ident,
+                                  ident: folder.fold_ident(f.node.ident),
                                   pat: folder.fold_pat(f.node.pat),
                                   is_shorthand: f.node.is_shorthand,
                               }}

--- a/src/test/run-pass/issue-38556.rs
+++ b/src/test/run-pass/issue-38556.rs
@@ -1,0 +1,22 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+pub struct Foo;
+
+macro_rules! reexport {
+    () => { use Foo as Bar; }
+}
+
+reexport!();
+
+fn main() {
+    use Bar;
+    fn f(_: Bar) {}
+}

